engine/client/cl_font.c

```diff
@@ -378,3 +409,178 @@ void CL_DrawStringLen( cl_font_t *font, const char *s, int *width, int *height,
 		s++;
 	}
 }
+
+/* Font Class */
+
+qboolean Font_Init(Font* self, const char* name, int tall)
+{
+	int len = 0;
+	int x0, y0, x1, y1;
+
+	if( !self )
+		return false;
+
+	self->m_bIsLoaded = false;
+
+	COM_FreeFile(self->m_pFontData);
+	self->m_pFontData = COM_LoadFileForMe( name, &len );
+	if (self->m_pFontData == NULL)
+		return false;
+
+	if (!stbtt_InitFont(&self->m_FontInfo, self->m_pFontData, 0))
+		return false;
+
+	self->scale = stbtt_ScaleForPixelHeightPrecision( &self->m_FontInfo, tall + 4 );
+
+	stbtt_GetFontVMetrics( &self->m_FontInfo, &self->m_iAscent, NULL, NULL );
+	self->m_iAscent = round( self->m_iAscent * self->scale );
+
+	stbtt_GetFontBoundingBox( &self->m_FontInfo, &x0, &y0, &x1, &y1 );
+	self->m_iHeight = round(( y1 - y0 ) * self->scale ); // maybe wrong!
+	self->m_iMaxCharWidth = round(( x1 - x0 ) * self->scale ); // maybe wrong!
+
+	self->m_bIsLoaded = true;
+
+	return true;
+}
+
+int Font_CheckCharExists(Font* self, int ch)
+{
+	for (int i = 0; i < self->m_iCharCount; i++)
+	{
+		if (Font_GetTexIndex(self, ch) == self->m_iBuffer[i])
+			return i;
+	}
+
+	return -1;
+}
+
+const int RGBA_Channels = 4; // RGBA Channels
+int Font_LoadChar(Font* self, int ch)
+{
+	byte *buf;
+
+	int id, bm_top, bm_left, bm_height, bm_width;
+	char texname[30];
+
+	int iCallBack = Font_CheckCharExists(self, ch);
+	if (iCallBack > -1) return iCallBack;
+
+	buf = stbtt_GetCodepointBitmap( &self->m_FontInfo, self->scale, self->scale, ch, &bm_width, &bm_height, &bm_left, &bm_top );
+
+	Q_snprintf( texname, sizeof( texname ), "font_%i", Font_GetTexIndex(self, ch));
+
+	if (!(id = ref.dllFuncs.GL_FindTexture(texname)))
```

Comment:
Something that would be really nice to have, before this gets merged, is to load the characters in the atlas texture, like mainui_cpp, to eliminate texture switching. 

Maybe allocate 1024x1024 texture and fill it with characters as needed, and once it gets filled up, allocate another large texture.

And then cache the result, something I did in `mainui_cpp` because drawing glyphs takes a lot of time.